internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,595 @@
+	}
+	optionalEntry = &authd.UILayout{
+		Type:  "optional-entry",
+		Entry: &optionalEntries,
+	}
+	emptyType = &authd.UILayout{
+		Type:  "",
+		Entry: &requiredEntries,
+	}
+)
+
+func TestAvailableBrokers(t *testing.T) {
+	t.Parallel()
+
+	abResp, err := testClient.AvailableBrokers(context.Background(), &authd.Empty{})
+	require.NoError(t, err, "AvailableBrokers should not return an error, but did")
+
+	got := abResp.GetBrokersInfos()
+	for _, broker := range got {
+		broker.Id = broker.Name + "_ID"
+	}
+	want := testutils.LoadWithUpdateFromGoldenYAML(t, got)
+	require.Equal(t, want, got, "AvailableBrokers returned unexpected brokers")
+}
+
+func TestGetPreviousBroker(t *testing.T) {
+	t.Parallel()
+
+	username := t.Name()
+
+	// Try to get the broker for the user before assigning it.
+	gotResp, _ := testClient.GetPreviousBroker(context.Background(), &authd.GPBRequest{Username: username})
+	require.Empty(t, gotResp.GetPreviousBroker(), "GetPreviousBroker should return nil when the user has no broker assigned")
+
+	_, err := testClient.SetDefaultBrokerForUser(context.Background(), &authd.SDBFURequest{
+		BrokerId: "local",
+		Username: username,
+	})
+	require.NoError(t, err, "Setup: could not set default broker for user for tests")
+
+	// Assert that the broker assigned to the user is correct.
+	gotResp, _ = testClient.GetPreviousBroker(context.Background(), &authd.GPBRequest{Username: username})
+	require.Equal(t, "local", gotResp.GetPreviousBroker(), "GetPreviousBroker did not return the correct broker")
+
+	// Asset that trying to get the broker from a user with no broker assigned returns empty.
+	gotResp, _ = testClient.GetPreviousBroker(context.Background(), &authd.GPBRequest{Username: username + "no broker"})
+	require.Empty(t, gotResp.GetPreviousBroker(), "GetPreviousBroker should return nil when the user has no broker assigned")
+}
+
+func TestSelectBroker(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		// These are the function arguments.
+		brokerID string
+		username string
+
+		// This is the expected return.
+		wantErr bool
+	}{
+		"Successfully select a broker and creates the session": {username: t.Name()},
+
+		"Error when username is empty":                    {wantErr: true},
+		"Error when brokerID is empty":                    {username: "empty broker", brokerID: "-", wantErr: true},
+		"Error when broker does not exist":                {username: "no broker", brokerID: "does not exist", wantErr: true},
+		"Error when broker does not provide a session ID": {username: "NS_no_id", wantErr: true},
+		"Error when starting the session":                 {username: "NS_error", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.brokerID == "" {
+				tc.brokerID = brokerGeneratedID
+			} else if tc.brokerID == "-" {
+				tc.brokerID = ""
+			}
+
+			sbRequest := &authd.SBRequest{
+				BrokerId: tc.brokerID,
+				Username: tc.username,
```

Comment:
for the fixed named, you are using here username that are shared (even if supposively only the error cases, but we never know)

Reply:
This is not a problem. If the NewSession fails, nothing gets assigned so there's no risk of conflicts with other tests, even if the username is the same.

Reply:
Yes! But you are relying on knowing the internal implementation. Let’s say there is a change or a bug that something is assigned. Hence it’s better to stay safe, does it make sense?